app/src/main/java/com/mertceyhan/bitcoinmarket/widget/ui/BitcoinWidget.kt

```diff
@@ -0,0 +1,124 @@
+package com.mertceyhan.bitcoinmarket.widget.ui
+
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import androidx.glance.*
+import androidx.glance.action.actionStartActivity
+import androidx.glance.action.clickable
+import androidx.glance.appwidget.action.actionRunCallback
+import androidx.glance.layout.*
+import androidx.glance.text.FontWeight
+import androidx.glance.text.Text
+import androidx.glance.text.TextStyle
+import androidx.glance.unit.ColorProvider
+import com.mertceyhan.R
+import com.mertceyhan.bitcoinmarket.features.MainActivity
+import com.mertceyhan.bitcoinmarket.widget.callback.MarketRefreshCallback
+
+@Composable
+fun BitcoinWidget(currentPrice: String?, changeRate: String?, isChangeRatePositive: Boolean?) {
+
+    Column(
+        modifier = GlanceModifier
+            .width(170.dp).height(130.dp)
+            .background(ImageProvider(R.drawable.background_widget))
+            .clickable(actionStartActivity(activity = MainActivity::class.java))
+            .padding(8.dp)
+    ) {
+        BitcoinWidgetHeader()
+        BitcoinWidgetBody(currentPrice)
+        BitcoinFooterBody(isChangeRatePositive ?: false, changeRate)
+    }
+}
+
+@Composable
+fun BitcoinWidgetHeader() {
+
+    Row(
+        modifier = GlanceModifier.fillMaxWidth()
+    ) {
+        Image(
+            provider = ImageProvider(R.drawable.ic_btc_flat),
+            contentDescription = "btc"
+        )
+        Row(
+            horizontalAlignment = Alignment.End,
+            modifier = GlanceModifier.fillMaxWidth()
+        ) {
+            Image(
+                provider = ImageProvider(R.drawable.ic_refresh),
+                contentDescription = "refresh",
+                modifier = GlanceModifier.clickable(actionRunCallback<MarketRefreshCallback>())
+            )
+        }
+    }
+}
+
+@Composable
+fun BitcoinWidgetBody(currentPrice: String?) {
+
+    val context = LocalContext.current
+
+    Text(
+        text = context.getString(R.string.bitcoin_label),
+        modifier = GlanceModifier.fillMaxWidth().padding(top = 8.dp),
+        style = TextStyle(fontWeight = FontWeight.Bold),
+    )
+    Text(
+        text = currentPrice ?: context.getString(R.string.fetch_alert_label),
+        modifier = GlanceModifier.fillMaxWidth().padding(top = 8.dp),
+        style = TextStyle(fontWeight = FontWeight.Bold, fontSize = 16.sp),
+    )
+}
+
+@Composable
+fun BitcoinFooterBody(isChangeRatePositive: Boolean, changeRate: String?) {
+
+    val context = LocalContext.current
+
+    val rateBackground = if (isChangeRatePositive) {
```

Comment:
Could you add this kind of UI logic into the UI model and write unit tests for it, please?

Reply:
Please check out `MarketScreenViewState` class for an example.